File: java/src/main/java/org/licenselynx/QuotesHandler.java
package org.licenselynx;

/**
 * Handler for quotes in input strings.
 */
final class QuotesHandler
{

    // Array of quote characters to be replaced.
    private static final char[] QUOTE_CHARACTERS = {
            // Single quotes
            '‘', // LEFT SINGLE QUOTATION MARK
            '’', // RIGHT SINGLE QUOTATION MARK
            '‚', // SINGLE LOW-9 QUOTATION MARK
            '‛', // SINGLE HIGH-REVERSED-9 QUOTATION MARK
            '′', // PRIME (often used as an apostrophe)
            '＇', // FULLWIDTH APOSTROPHE
            // Double quotes
            '“', // LEFT DOUBLE QUOTATION MARK
            '”', // RIGHT DOUBLE QUOTATION MARK
            '„', // DOUBLE LOW-9 QUOTATION MARK
            '‟', // DOUBLE HIGH-REVERSED-9 QUOTATION MARK
            '″', // DOUBLE PRIME
            '«', // LEFT-POINTING DOUBLE ANGLE QUOTATION MARK
            '»', // RIGHT-POINTING DOUBLE ANGLE QUOTATION MARK
            '＂'  // FULLWIDTH QUOTATION MARK
    };

    // Private constructor prevents instantiation.
    private QuotesHandler()
    {
        throw new AssertionError("Utility class should not be instantiated.");
    }

    /**
     * Normalizes the input string by replacing various Unicode quote characters
     * with the default replacement (') character.
     *
     * @param pInputString The input text that may contain various quote characters.
     * @return The normalized text with all recognized quote characters replaced.
     */
    public static String normalizeQuotes(final String pInputString)
    {
        return normalizeQuotes(pInputString, "'");
    }

    /**
     * Normalizes the input string by replacing various Unicode quote characters
     * with a specified replacement string.
     *
     * @param pInputString The input text that may contain various quote characters.
     * @param pReplacement The quote character replacement to use.
     * @return The normalized text with all recognized quote characters replaced.
     */
    public static String normalizeQuotes(final String pInputString, final String pReplacement)
    {
        if (pInputString == null)
        {
            return null;
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < pInputString.length(); i++)
        {
            char c = pInputString.charAt(i);
            if (isQuoteCharacter(c))
            {
                sb.append(pReplacement);
            }
            else {
                sb.append(c);
            }
        }
        return sb.toString();
    }

    /**
     * Checks whether the given character is in the defined list of quote characters.
     *
     * @param pCharacter The character to check.
     * @return True if the character is a quote character, false otherwise.
     */
    private static boolean isQuoteCharacter(final char pCharacter)
    {
        for (char quote : QUOTE_CHARACTERS)
        {
            if (quote == pCharacter)
            {
                return true;
            }
        }
        return false;
    }
}

File: data/licenselynx-licenselynx/python/licenselynx/quotes_handler.py
#
# SPDX-FileCopyrightText: Copyright 2025 Siemens AG
# SPDX-License-Identifier: BSD-3-Clause
#
class _QuotesHandler:
    # List of quote characters to replace with the default replacement.
    quote_characters = [
        # Single quotes
        "\u2018",  # LEFT SINGLE QUOTATION MARK ‘
        "\u2019",  # RIGHT SINGLE QUOTATION MARK ’
        "\u201A",  # SINGLE LOW-9 QUOTATION MARK ‚
        "\u201B",  # SINGLE HIGH-REVERSED-9 QUOTATION MARK ‛
        "\u2032",  # PRIME (often used as an apostrophe) ′
        "\uFF07",  # FULLWIDTH APOSTROPHE ＇
        # Double quotes
        "\u201C",  # LEFT DOUBLE QUOTATION MARK “
        "\u201D",  # RIGHT DOUBLE QUOTATION MARK ”
        "\u201E",  # DOUBLE LOW-9 QUOTATION MARK „
        "\u201F",  # DOUBLE HIGH-REVERSED-9 QUOTATION MARK ‟
        "\u2033",  # DOUBLE PRIME ″
        "\u00AB",  # LEFT-POINTING DOUBLE ANGLE QUOTATION MARK «
        "\u00BB",  # RIGHT-POINTING DOUBLE ANGLE QUOTATION MARK »
        "\uFF02",  # FULLWIDTH QUOTATION MARK ＂
    ]

    def normalize_quotes(self, input_string: str, replacement: str = "'") -> str:
        """
        Normalize a string by replacing various Unicode and ASCII quote characters
        with a single default quote character. By default, all quotes (both single and double)
        are mapped to the ASCII single quote.

        Args:
            input_string (str): The input text that potentially contains various quote characters.
            replacement (str, optional): The quote character replacement to use. Defaults to "'".

        Returns:
            str: The normalized text with all recognized quote characters replaced.
        """
        translation_map = {ord(char): replacement for char in self.quote_characters}
        normalized_input_string = input_string.translate(translation_map)
        return normalized_input_string
